core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -88,7 +91,18 @@ private ParallelIterator(
     @Override
     public void close() {
       // close first, avoid new task submit
-      this.closed = true;
+      this.closed.set(true);
+
+      for (Task<T> task : yieldedTasks) {
+        try {
+          task.close();
+        } catch (Exception e) {
+          throw new RuntimeException("Close failed", e);
```

Comment:
we may want to finish the close for loop in case of failure in the middle. should we use the `Tasks` util here?

Here is an example from `CatalogUtil::deleteFile`
```
Tasks.foreach(files)
        .executeWith(ThreadPools.getWorkerPool())
        .noRetry()
        .suppressFailureWhenFinished()
        .onFailure((file, exc) -> LOG.warn("Failed to delete {} file {}", type, file, exc))
        .run(io::deleteFile);
```

Reply:
> we may want to finish the close for loop in case of failure in the middle.

good point, thanks!
added `reliably attempt close on all yielded tasks` commit